File: cpp/101-150/126. Word Ladder II.cpp
class Solution {
public:
    void dfshelper(string word, vector<vector<string>> &res, vector<string> &path, map<string, bool> &wordDict, string endWord) {
        if (!wordDict[word] || find(path.begin(), path.end(), word) != -1)
            return;
        if (word == endWord) {
            res.push_back(path);
            return;
        }
        for (int i = 0; i < word.size(); i++) {
            string newWord = word;
            for (int j = 0; j < 26; j++) {
            	newWord[i] = 'a' + j;
            	if(newWord == word)
            		continue;
                path.push_back(newWord);
                dfshelper(newWord, res, path, wordDict, endWord);
                path.pop_back();
            }
        }
    }
    vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
        vector<vector<string >> res;
        map<string, bool> wordDict;
        for (auto word : wordList) {
            wordDict[word] = true;
        }
        vector<string> path;
        for (int i = 0; i < beginWord.size(); i++) {
            string newWord = beginWord;
            for (int j = 0; j < 26; j++) {
            	newWord[i] = 'a' + j;
            	if(newWord == beginWord)
            		continue;
                path.push_back(newWord);
                dfshelper(newWord, res, path, wordDict, endWord);
                path.pop_back();
            }
        }
        return res;
    }

};

File: python/101-150/126. Word Ladder II.py


class Solution(object):

    def getnext(self, s):
        res = []
        for i in range(len(s)):
            for c in 'abcdefghijklmnopqrstuvwxyz':
                if c != s[i]:
                    res.append(s[:i] + c + s[i + 1:])
        return res

    def helper(self, path, res, endWord):
        if path[-1] == endWord:
            res.append(path)
            return
        for word in self.getnext(path[-1]):
            if word in self.distance and self.distance[word] - 1 == self.distance[path[-1]]:
                self.helper(path + [word], res, endWord)

    def findLadders(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: List[List[str]]
        """
        worddic = {}
        for word in wordList:
            worddic[word] = True
        if endWord not in wordList:
            return []
        level = [beginWord]
        self.distance = {}
        self.distance[beginWord] = 0
        while level:
            temp = []
            for word1 in level:
                for word2 in self.getnext(word1):
                    if word2 in worddic and word2 not in self.distance:
                        self.distance[word2] = self.distance[word1] + 1
                        temp.append(word2)
            level = temp
        res = []
        self.helper([beginWord], res, endWord)
        return res

sol = Solution()
begin = "hit"
end = "cog"
list = ["hot","dot","dog","lot","log","cog"]
print sol.findLadders(begin, end, list)
